force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,15 +244,21 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+        //Intialize a list to hold contacts to be upserted
+        List <Contact> contactsToUpsertList = new List <Contact>();
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
+        for (Contact contact : contactsToUpsert) { 
+            // Extract the Account Name from Contact's LastName
+            String accountName = contact.LastName;
+            // Get the account based on the name or create one if it doesn't exist
+            Account account =  upsertAccount (accountName);
```

Comment:
This works well to accomplish the goal of associating the Contacts with matching Accounts, but for an extra challenge, try refactoring this to not have DML/SOQL inside of a for loop.
Hint: Follow these steps

1. Collect unique last names from the list of Contacts to upsert
2. Query existing Accounts with names matching the last names
3. Create new Accounts for last names that do not have a matching Account
4. Associate Contacts with their corresponding Accounts
5. Upsert the Contacts